src/robin_dict.jl

```diff
@@ -0,0 +1,593 @@
+Keys are compared with [`isequal`](@ref) and hashed with [`hash`](@ref).
+Given a single iterable argument, constructs a [`RobinDict`](@ref) whose key-value pairs
+are taken from 2-tuples `(key,value)` generated by the argument.
+
+# Examples
+```jldoctest
+julia> RobinDict([("A", 1), ("B", 2)])
+RobinDict{String,Int64} with 2 entries:
+  "B" => 2
+  "A" => 1
+```
+
+Alternatively, a sequence of pair arguments may be passed.
+
+```jldoctest
+julia> RobinDict("A"=>1, "B"=>2)
+RobinDict{String,Int64} with 2 entries:
+  "B" => 2
+  "A" => 1
+```
+"""
+mutable struct RobinDict{K,V} <: AbstractDict{K,V}
+    #there is no need to maintain an table_size as an additional variable
+    # dibs::Array{Int8,1} # distance to initial bucket - critical for implementation
+    hashes::Vector{UInt32}
+    keys::Array{K,1}
+    vals::Array{V,1}
+    count::Int
+    totalcost::Int
+    maxprobe::Int   # length of longest probe
+    idxfloor::Int
+
+    function RobinDict{K, V}() where {K, V}
+        n = 16 # default size of an empty Dict in Julia
+        new(zeros(UInt32, n), Vector{K}(undef, n), Vector{V}(undef, n), 0, 0, 0, 0)
+    end
+
+    function RobinDict{K, V}(d::RobinDict{K, V}) where {K, V}
+        new(copy(d.hashes), copy(d.keys), copy(d.vals), d.count, d.totalcost, d.maxprobe, d.idxfloor)
+    end
+
+    function RobinDict{K, V}(keys, vals, hashes, count, totalcost, maxprobe, idxfloor) where {K, V}
+        new(hashes, keys, vals, count, totalcost, maxprobe, idxfloor)
+    end
+end
+
+function RobinDict{K,V}(kv) where V where K
+    h = RobinDict{K,V}()
+    for (k,v) in kv
+        h[k] = v
+    end
+    return h
+end
+RobinDict{K,V}(p::Pair) where {K,V} = setindex!(RobinDict{K,V}(), p.second, p.first)
+function RobinDict{K,V}(ps::Pair...) where V where K
+    h = RobinDict{K,V}()
+    sizehint!(h, length(ps))
+    for p in ps
+        h[p.first] = p.second
+    end
+    return h
+end
+
+RobinDict() = RobinDict{Any,Any}()
+RobinDict(kv::Tuple{}) = RobinDict()
+copy(d::RobinDict) = RobinDict(d)
+
+RobinDict(ps::Pair{K,V}...) where {K,V} = RobinDict{K,V}(ps)
+RobinDict(ps::Pair...)                  = RobinDict(ps)
+
+function RobinDict(d::AbstractDict{K, V}) where {K, V}
+    h = RobinDict{K, V}()
+    for (k, v) in d
+        h[k] = v
+    end
+    h
+end
+
+function RobinDict(kv)
+    try
+        d = dict_with_eltype((K, V) -> RobinDict{K, V}, kv, eltype(kv))
+        if isa(d, Dict)
+            return RobinDict(d)
+        else
+            return d
+        end
+    catch e
+    if !isiterable(typeof(kv)) || !all(x -> isa(x, Union{Tuple,Pair}), kv)
+            !all(x->isa(x,Union{Tuple,Pair}),kv)
+            throw(ArgumentError("RobinDict(kv): kv needs to be an iterator of tuples or pairs"))
+        else
+            rethrow(e)
+        end
+    end
+end
+
+hash_key(key) = ((hash(key)%UInt32) & 0xffffffff)|1
+desired_index(hash, sz) = ((hash) & (sz -1)) + 1
```

Comment:
The LSB of `hash_key` is always set and is always used in the `desired_index`. In other words, it looks like every entry wants to be in an even slot.

Is this intended?

Otherwise, one could try something like storing some `ashift` in the RobinDict such that `sz = 1<<(32 - ashift)`, and `desired_index(hash, ashift) = 1+(hash >>ashift)`. 

This would have the advantage that when we iterate over one RobinDict and use that to access another RobinDict, we get very favorable memory access patterns (almost linear), regardless of whether sizes match. This especially applies for `rehash!`. With the "reduction-by-masking", we get an interleaving of `new_size/old_size` many linear patterns.

If we stay with masking, one could store `hash_key(key) = (hash(key)%UInt32) | 0x80000000`

Reply:
> Is this intended?

No, it wasn't. Just did it for some reason